### PR TITLE
thrift_proxy: Add thrift header to metadata filter

### LINKED ISSUE #18578: support header-to-metadata filter for thrift
*Title*: *One line description*
support header-to-metadata filter for thrift

*Description*:
Is there any plan to support header-to-metadata filter for thrift ?
As http did

[optional *Relevant Links*:]
https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto

### PR DESCRIPTION
**Pull request by fishcakez**
Commit Message: thrift_proxy: Add thrift header to metadata filter
Additional Description: Transforming headers to metadata and then using that metadata for subset load balancing, logging or other custom functionality is fundamental part of envoy proxy capabilities. Therefore add header to metadata proto configuration that is exactly like the http equivalent.
Risk Level: low
Testing: N/A
Docs Changes: N/A
Release Notes: new feature
Platform Specific Features: N/A
Fixes #18578: Filter to support this feature.
API Considerations(https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md): Given this is a clone of the http header to metadata config (without cookie support and with the future one of promotion) I believe it should pass the checklist.

### CODE REVIEW

**Comment by mattklein123** [comment]
cc @rgs1 

/wait-any

**Review by rgs1 — DISMISSED** [review]
LGTM -- thanks!

cc: @envoyproxy/api-shepherds

**Comment by fishcakez** [comment]
https://storage.googleapis.com/envoy-pr/2d5b24d/coverage/source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc.gcov.html

looks like need add couple tests for two errors cases to get coverage up to scratch

**Comment by fishcakez** [comment]
ASAN failure is due to publishing test results timeout:
```
##[error]The HTTP request timed out after 00:00:30.
```

**Comment by mattklein123** [comment]
@rgs1 to do first pass review, thank you.

File: source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/config.cc [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+ThriftProxy::ThriftFilters::FilterFactoryCb
+HeaderToMetadataFilterConfig::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::network::thrift_proxy::filters::header_to_metadata::v3::
+        HeaderToMetadata& proto_config,
+    const std::string&, Server::Configuration::FactoryContext&) {
+  ConfigSharedPtr filter_config(std::make_shared<Config>(proto_config));
+  return
+      [filter_config](ThriftProxy::ThriftFilters::FilterChainFactoryCallbacks& callbacks) -> void {
+        callbacks.addDecoderFilter(std::make_shared<HeaderToMetadataFilter>(filter_config));
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/config.cc:22]: nit: we don't need a new shared ptr on every call, pass a ref to filter config?

File: source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+  }
+  // Any matching rules?
+  if (!structs_by_namespace.empty()) {
+    for (auto const& entry : structs_by_namespace) {
+      callbacks.streamInfo().setDynamicMetadata(entry.first, entry.second);
+    }
+  }
+}
+
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc:180]: Can you check if we handle #11747?

**Review by rgs1 — CHANGES_REQUESTED** [review]
2 comments - thanks!

**Review by xiaoma2015 — APPROVED** [review]

File: source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc (hunk 2 above)
Comment (fishcakez) [reply source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc:180]: I copied the test and it passes :)

**Review by rgs1 — APPROVED** [review]
LGTM - thanks!

Over to @zuercher or @mattklein123 for merging.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**